# Olist Seller Evaluation & Channel Analysis


# Part 0 — Data Inspection


In [79]:
import pandas as pd
import os

In [80]:
os.listdir()

['.ipynb_checkpoints',
 '01_data_inspection.ipynb',
 '02_data_cleaning.ipynb',
 'olist_closed_deals_dataset.csv',
 'olist_customers_dataset.csv',
 'olist_geolocation_dataset.csv',
 'olist_marketing_qualified_leads_dataset.csv',
 'olist_orders_dataset.csv',
 'olist_order_items_dataset.csv',
 'olist_order_payments_dataset.csv',
 'olist_order_reviews_dataset.csv',
 'olist_products_dataset.csv',
 'olist_sellers_dataset.csv',
 'product_category_name_translation.csv']

In [3]:
mql = pd.read_csv("olist_marketing_qualified_leads_dataset.csv")
closed = pd.read_csv("olist_closed_deals_dataset.csv")

In [4]:
print("MQL shape:", mql.shape)
print("Closed Deals shape:", closed.shape)

display(mql.head())
display(closed.head())

MQL shape: (8000, 4)
Closed Deals shape: (842, 14)


,mql_id,first_contact_date,landing_page_id,origin
0,dac32acd4db4c29c230538b72f8dd87d,2018-02-01,88740e65d5d6b056e0cda098e1ea6313,social
1,8c18d1de7f67e60dbd64e3c07d7e9d5d,2017-10-20,007f9098284a86ee80ddeb25d53e0af8,paid_search
2,b4bc852d233dfefc5131f593b538befa,2018-03-22,a7982125ff7aa3b2054c6e44f9d28522,organic_search
3,6be030b81c75970747525b843c1ef4f8,2018-01-22,d45d558f0daeecf3cccdffe3c59684aa,email
4,5420aad7fec3549a85876ba1c529bd84,2018-02-21,b48ec5f3b04e9068441002a19df93c6c,organic_search


,mql_id,seller_id,sdr_id,sr_id,won_date,business_segment,lead_type,lead_behaviour_profile,has_company,has_gtin,average_stock,business_type,declared_product_catalog_size,declared_monthly_revenue
0,5420aad7fec3549a85876ba1c529bd84,2c43fb513632d29b3b58df74816f1b06,a8387c01a09e99ce014107505b92388c,4ef15afb4b2723d8f3d81e51ec7afefe,2018-02-26 19:58:54,pet,online_medium,cat,NaN,NaN,NaN,reseller,NaN,0.0
1,a555fb36b9368110ede0f043dfc3b9a0,bbb7d7893a450660432ea6652310ebb7,09285259593c61296eef10c734121d5b,d3d1e91a157ea7f90548eef82f1955e3,2018-05-08 20:17:59,car_accessories,industry,eagle,NaN,NaN,NaN,reseller,NaN,0.0
2,327174d3648a2d047e8940d7d15204ca,612170e34b97004b3ba37eae81836b4c,b90f87164b5f8c2cfa5c8572834dbe3f,6565aa9ce3178a5caf6171827af3a9ba,2018-06-05 17:27:23,home_appliances,online_big,cat,NaN,NaN,NaN,reseller,NaN,0.0
3,f5fee8f7da74f4887f5bcae2bafb6dd6,21e1781e36faf92725dde4730a88ca0f,56bf83c4bb35763a51c2baab501b4c67,d3d1e91a157ea7f90548eef82f1955e3,2018-01-17 13:51:03,food_drink,online_small,NaN,NaN,NaN,NaN,reseller,NaN,0.0
4,ffe640179b554e295c167a2f6be528e0,ed8cb7b190ceb6067227478e48cf8dde,4b339f9567d060bcea4f5136b9f5949e,d3d1e91a157ea7f90548eef82f1955e3,2018-07-03 20:17:45,home_appliances,industry,wolf,NaN,NaN,NaN,manufacturer,NaN,0.0


In [5]:
print("MQL columns:")
print(mql.columns.tolist())

print("\nClosed Deals columns:")
print(closed.columns.tolist())

MQL columns:
['mql_id', 'first_contact_date', 'landing_page_id', 'origin']

Closed Deals columns:
['mql_id', 'seller_id', 'sdr_id', 'sr_id', 'won_date', 'business_segment', 'lead_type', 'lead_behaviour_profile', 'has_company', 'has_gtin', 'average_stock', 'business_type', 'declared_product_catalog_size', 'declared_monthly_revenue']


In [6]:
mql.info()
print("\n" + "="*60 + "\n")
closed.info()

<class 'pandas.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype
---  ------              --------------  -----
 0   mql_id              8000 non-null   str  
 1   first_contact_date  8000 non-null   str  
 2   landing_page_id     8000 non-null   str  
 3   origin              7940 non-null   str  
dtypes: str(4)
memory usage: 907.1 KB


<class 'pandas.DataFrame'>
RangeIndex: 842 entries, 0 to 841
Data columns (total 14 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   mql_id                         842 non-null    str    
 1   seller_id                      842 non-null    str    
 2   sdr_id                         842 non-null    str    
 3   sr_id                          842 non-null    str    
 4   won_date                       842 non-null    str    
 5   business_segment               841 non-null    str    
 6   lead

In [7]:
print("MQL missing values:")
display(mql.isna().sum().sort_values(ascending=False))

print("\nClosed Deals missing values:")
display(closed.isna().sum().sort_values(ascending=False))

MQL missing values:


origin                60
mql_id                 0
first_contact_date     0
landing_page_id        0
dtype: int64


Closed Deals missing values:


has_company                      779
has_gtin                         778
average_stock                    776
declared_product_catalog_size    773
lead_behaviour_profile           177
business_type                     10
lead_type                          6
business_segment                   1
won_date                           0
seller_id                          0
mql_id                             0
sr_id                              0
sdr_id                             0
declared_monthly_revenue           0
dtype: int64

In [8]:
print("MQL duplicate mql_id:", mql["mql_id"].duplicated().sum())
print("Closed duplicate mql_id:", closed["mql_id"].duplicated().sum())
print("Closed duplicate seller_id:", closed["seller_id"].duplicated().sum())

MQL duplicate mql_id: 0
Closed duplicate mql_id: 0
Closed duplicate seller_id: 0


In [9]:
print("Acquisition channels:")
display(mql["origin"].value_counts(dropna=False))

print("\nBusiness types:")
display(closed["business_type"].value_counts(dropna=False))

print("\nLead types:")
display(closed["lead_type"].value_counts(dropna=False))

print("\nBusiness segments:")
display(closed["business_segment"].value_counts(dropna=False))

Acquisition channels:


origin
organic_search       2296
paid_search          1586
social               1350
unknown              1099
direct_traffic        499
email                 493
referral              284
other                 150
display               118
other_publicities      65
NaN                    60
Name: count, dtype: int64


Business types:


business_type
reseller        587
manufacturer    242
NaN              10
other             3
Name: count, dtype: int64


Lead types:


lead_type
online_medium      332
online_big         126
industry           123
offline            104
online_small        77
online_beginner     57
online_top          14
NaN                  6
other                3
Name: count, dtype: int64


Business segments:


business_segment
home_decor                         105
health_beauty                       93
car_accessories                     77
household_utilities                 71
construction_tools_house_garden     69
audio_video_electronics             64
computers                           34
pet                                 30
food_supplement                     28
food_drink                          26
sports_leisure                      25
bed_bath_table                      22
bags_backpacks                      22
toys                                20
fashion_accessories                 19
home_office_furniture               14
stationery                          13
phone_mobile                        13
small_appliances                    12
handcrafted                         12
baby                                10
music_instruments                    9
books                                9
watches                              8
jewerly                              8
home_app

In [10]:
print("Unique SDRs:", closed["sdr_id"].nunique())
print("Unique SRs:", closed["sr_id"].nunique())
print("Unique sellers:", closed["seller_id"].nunique())

Unique SDRs: 32
Unique SRs: 22
Unique sellers: 842


# INITIAL EXPLORATORY ANALYSIS
# Preliminary observations only — before cleaning and validation

In [11]:
funnel_initial = mql.merge(closed,on="mql_id",how="left")

In [12]:
print(mql.shape)
print(closed.shape)
print(funnel_initial.shape)

(8000, 4)
(842, 14)
(8000, 17)


In [13]:
funnel_initial["converted"] = funnel_initial["seller_id"].notna()

In [14]:
funnel_initial["converted"].value_counts()

converted
False    7158
True      842
Name: count, dtype: int64

In [15]:
conversion_rate_initial = funnel_initial["converted"].mean() * 100

In [16]:
conversion_rate_initial

np.float64(10.525)

In [17]:
print("Initial conversion rate:", round(conversion_rate_initial, 2), "%")

Initial conversion rate: 10.52 %


In [18]:
#Which marketing channels seem to convert leads into sellers more often?


In [19]:
mql["origin"].value_counts()

origin
organic_search       2296
paid_search          1586
social               1350
unknown              1099
direct_traffic        499
email                 493
referral              284
other                 150
display               118
other_publicities      65
Name: count, dtype: int64

In [20]:
funnel_initial[funnel_initial["converted"] == True]["origin"].value_counts()

origin
organic_search       271
paid_search          195
unknown              179
social                75
direct_traffic        56
referral              24
email                 15
display                6
other                  4
other_publicities      3
Name: count, dtype: int64

In [21]:
total_leads_by_channel = mql["origin"].value_counts()

converted_by_channel = funnel_initial[funnel_initial["converted"] == True]["origin"].value_counts()

In [22]:
channel_conversion_rate = (converted_by_channel / total_leads_by_channel) * 100

In [23]:
channel_conversion_rate.sort_values(ascending=False)

origin
unknown              16.287534
paid_search          12.295082
organic_search       11.803136
direct_traffic       11.222445
referral              8.450704
social                5.555556
display               5.084746
other_publicities     4.615385
email                 3.042596
other                 2.666667
Name: count, dtype: float64

Preliminary observation: Lead-to-seller conversion differs across acquisition channels. paid_search, organic_search, and direct_traffic currently show conversion rates above the overall preliminary rate of 10.52%. The unknown category has the highest observed rate, but it cannot yet be interpreted as an actionable acquisition channel. These findings are exploratory and will be validated after data cleaning and seller-performance analysis.

In [81]:
orders = pd.read_csv("olist_orders_dataset.csv")
order_items = pd.read_csv("olist_order_items_dataset.csv")
sellers = pd.read_csv("olist_sellers_dataset.csv")
reviews = pd.read_csv("olist_order_reviews_dataset.csv")

In [36]:
print("Orders:", orders.shape)
print("Order Items:", order_items.shape)
print("Sellers:", sellers.shape)
print("Reviews:", reviews.shape)

Orders: (99441, 8)
Order Items: (112650, 7)
Sellers: (3095, 4)
Reviews: (99224, 7)


In [37]:
print("Orders columns:")
print(orders.columns.tolist())

Orders columns:
['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']


In [38]:
print("\nOrder Items columns:")
print(order_items.columns.tolist())


Order Items columns:
['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']


In [39]:
print("\nSellers columns:")
print(sellers.columns.tolist())


Sellers columns:
['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state']


In [40]:
print("\nReviews columns:")
print(reviews.columns.tolist())


Reviews columns:
['review_id', 'order_id', 'review_score', 'review_comment_title', 'review_comment_message', 'review_creation_date', 'review_answer_timestamp']


In [41]:
display(orders.head())

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [42]:
display(order_items.head())

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [43]:
display(sellers.head())

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [44]:
display(reviews.head())

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


In [48]:
customers = pd.read_csv("olist_customers_dataset.csv")

print("Customers shape:", customers.shape)
print(customers.columns.tolist())

display(customers.head())

Customers shape: (99441, 5)
['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [49]:
print("Orders duplicate order_id:",
      orders["order_id"].duplicated().sum())

Orders duplicate order_id: 0


In [50]:
print("Order Items duplicate order_id:",
      order_items["order_id"].duplicated().sum())

Order Items duplicate order_id: 13984


In [51]:
print("Sellers duplicate seller_id:",
      sellers["seller_id"].duplicated().sum())

Sellers duplicate seller_id: 0


In [52]:
print("Customers duplicate customer_id:",
      customers["customer_id"].duplicated().sum())

Customers duplicate customer_id: 0


In [53]:
print("Reviews duplicate order_id:",
      reviews["order_id"].duplicated().sum())

Reviews duplicate order_id: 551


In [54]:
print("Unique orders:", orders["order_id"].nunique())

Unique orders: 99441


In [55]:
print("Unique sellers in sellers table:", sellers["seller_id"].nunique())

Unique sellers in sellers table: 3095


In [56]:
print("Unique sellers in order items:", order_items["seller_id"].nunique())

Unique sellers in order items: 3095


In [57]:
print("Unique customers:", customers["customer_unique_id"].nunique())

Unique customers: 96096


In [58]:
#missing values 
print("ORDERS")
print(orders.isna().sum())

print("\nORDER ITEMS")
print(order_items.isna().sum())

print("\nSELLERS")
print(sellers.isna().sum())

print("\nCUSTOMERS")
print(customers.isna().sum())

print("\nREVIEWS")
print(reviews.isna().sum())

ORDERS
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

ORDER ITEMS
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

SELLERS
seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

CUSTOMERS
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

REVIEWS
review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message    

In [59]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [60]:
reviews["review_score"].value_counts().sort_index()

review_score
1    11424
2     3151
3     8179
4    19142
5    57328
Name: count, dtype: int64

In [61]:
reviews["review_score"].mean()

np.float64(4.08642062404257)

In [62]:
order_items["price"].describe()

count    112650.000000
mean        120.653739
std         183.633928
min           0.850000
25%          39.900000
50%          74.990000
75%         134.900000
max        6735.000000
Name: price, dtype: float64

In [63]:
order_items["freight_value"].describe()

count    112650.000000
mean         19.990320
std          15.806405
min           0.000000
25%          13.080000
50%          16.260000
75%          21.150000
max         409.680000
Name: freight_value, dtype: float64

In [64]:
sellers["seller_state"].value_counts()

seller_state
SP    1849
PR     349
MG     244
SC     190
RJ     171
RS     129
GO      40
DF      30
ES      23
BA      19
CE      13
PE       9
PB       6
RN       5
MS       5
MT       4
RO       2
SE       2
AC       1
PI       1
MA       1
AM       1
PA       1
Name: count, dtype: int64

In [65]:
sellers["seller_city"].value_counts().head(10)

seller_city
sao paulo         694
curitiba          127
rio de janeiro     96
belo horizonte     68
ribeirao preto     52
guarulhos          50
ibitinga           49
santo andre        45
campinas           41
maringa            40
Name: count, dtype: int64

In [66]:
customers["customer_state"].value_counts()

customer_state
SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
PE     1652
CE     1336
PA      975
MT      907
MA      747
MS      715
PB      536
PI      495
RN      485
AL      413
SE      350
TO      280
RO      253
AM      148
AC       81
AP       68
RR       46
Name: count, dtype: int64

In [67]:
customers["customer_city"].value_counts().head(10)

customer_city
sao paulo                15540
rio de janeiro            6882
belo horizonte            2773
brasilia                  2131
curitiba                  1521
campinas                  1444
porto alegre              1379
salvador                  1245
guarulhos                 1189
sao bernardo do campo      938
Name: count, dtype: int64

In [68]:
orders["order_purchase_timestamp"] = pd.to_datetime(orders["order_purchase_timestamp"])

In [69]:
orders["order_delivered_customer_date"] = pd.to_datetime(orders["order_delivered_customer_date"])

In [70]:
orders["order_estimated_delivery_date"] = pd.to_datetime(orders["order_estimated_delivery_date"])

In [71]:
print("First order:",orders["order_purchase_timestamp"].min())

First order: 2016-09-04 21:15:19


In [72]:
print("Last order:",orders["order_purchase_timestamp"].max())

Last order: 2018-10-17 17:30:18


In [73]:
orders["order_purchase_timestamp"].dt.year.value_counts().sort_index()

order_purchase_timestamp
2016      329
2017    45101
2018    54011
Name: count, dtype: int64

In [74]:
orders["delivery_days_initial"] = (orders["order_delivered_customer_date"]- orders["order_purchase_timestamp"]).dt.days

In [75]:
orders["delivery_days_initial"].describe()

count    96476.000000
mean        12.094086
std          9.551746
min          0.000000
25%          6.000000
50%         10.000000
75%         15.000000
max        209.000000
Name: delivery_days_initial, dtype: float64

In [76]:
closed["seller_id"].isin(order_items["seller_id"]).value_counts()

seller_id
False    462
True     380
Name: count, dtype: int64

In [77]:
closed["seller_id"].isin(order_items["seller_id"]).mean() * 100

np.float64(45.13064133016627)

In [78]:
###Preliminary only — before cleaning and validation

# Part 1 — Data Cleaning


In [ ]:
import pandas as pd

In [ ]:
mql = pd.read_csv("olist_marketing_qualified_leads_dataset.csv")
closed = pd.read_csv("olist_closed_deals_dataset.csv")
orders = pd.read_csv("olist_orders_dataset.csv")
order_items = pd.read_csv("olist_order_items_dataset.csv")
sellers = pd.read_csv("olist_sellers_dataset.csv")
customers = pd.read_csv("olist_customers_dataset.csv")
reviews = pd.read_csv("olist_order_reviews_dataset.csv")

In [ ]:
print("MQL exact duplicates:", mql.duplicated().sum())
print("Closed exact duplicates:", closed.duplicated().sum())
print("Orders exact duplicates:", orders.duplicated().sum())
print("Order Items exact duplicates:", order_items.duplicated().sum())
print("Sellers exact duplicates:", sellers.duplicated().sum())
print("Customers exact duplicates:", customers.duplicated().sum())
print("Reviews exact duplicates:", reviews.duplicated().sum())

In [ ]:
print("Duplicate review order IDs:",reviews["order_id"].duplicated().sum())

In [ ]:
print("Duplicate order_id + order_item_id:",order_items[["order_id", "order_item_id"]].duplicated().sum())


In [ ]:
mql["first_contact_date"] = pd.to_datetime(mql["first_contact_date"])
closed["won_date"] = pd.to_datetime(closed["won_date"])
orders["order_purchase_timestamp"] = pd.to_datetime(orders["order_purchase_timestamp"])
orders["order_delivered_customer_date"] = pd.to_datetime(orders["order_delivered_customer_date"])
orders["order_estimated_delivery_date"] = pd.to_datetime(orders["order_estimated_delivery_date"])

In [ ]:
orders["order_status"] == "delivered"

In [ ]:
review_by_order = (reviews.groupby("order_id")["review_score"].mean().reset_index())

In [ ]:
#more than 1  id bec review = 1=2  2=8
review_by_order

In [ ]:
print(review_by_order.shape)
print("Duplicate order IDs:", review_by_order["order_id"].duplicated().sum())

In [ ]:
delivered_orders = orders[orders["order_status"] == "delivered"].copy()

In [ ]:
delivered_orders

In [ ]:
#removing rows without date of delivery 
delivered_orders = delivered_orders.dropna(subset=["order_delivered_customer_date"]).copy()

In [ ]:
delivered_orders["delivery_days"] = (delivered_orders["order_delivered_customer_date"]- delivered_orders["order_purchase_timestamp"]).dt.days

In [ ]:
delivered_orders["delivery_days"].describe()

In [ ]:
#build and validate seller orders

In [ ]:
seller_order_level_all = (order_items.groupby(["seller_id", "order_id"]).agg(order_value=("price", "sum"),freight_value=("freight_value", "sum"),item_count=("order_item_id", "count")).reset_index())


In [ ]:
print("Seller-order rows:",seller_order_level_all.shape)

print("Duplicate seller-order combinations:",seller_order_level_all[["seller_id", "order_id"]].duplicated().sum())

In [ ]:
seller_order_level_all = seller_order_level_all.merge(orders[["order_id","customer_id","order_status","order_purchase_timestamp","order_delivered_customer_date","order_estimated_delivery_date"]],on="order_id",how="left")
#seller-order rows left

In [ ]:
print("Seller-order rows after orders merge:",len(seller_order_level_all))
print("Missing order status:",seller_order_level_all["order_status"].isna().sum())

In [ ]:
seller_order_level_all = seller_order_level_all.merge(review_by_order,on="order_id",how="left")

In [ ]:
print("Rows after review merge:",len(seller_order_level_all))

print("Missing review scores:",seller_order_level_all["review_score"].isna().sum())

In [ ]:
seller_order_level_all = seller_order_level_all.merge(sellers,on="seller_id",how="left")

In [ ]:
print("Rows after seller merge:",len(seller_order_level_all))

print("Missing seller state:",seller_order_level_all["seller_state"].isna().sum())

print("Missing seller city:",seller_order_level_all["seller_city"].isna().sum())

In [ ]:
seller_order_level_delivered = seller_order_level_all[seller_order_level_all["order_status"] == "delivered"].copy()
#كم order لكل seller
# sales لكل seller
#review المرتبط بطلباته
# seller موقع

In [ ]:
print("Delivered seller-order rows:",len(seller_order_level_delivered))

print("Unique delivered sellers:",seller_order_level_delivered["seller_id"].nunique())

In [ ]:
seller_summary_delivered = (seller_order_level_delivered.groupby("seller_id").agg(total_orders=("order_id", "nunique"),total_revenue=("order_value", "sum"),average_review_score=("review_score", "mean"),review_count=("review_score", "count")).reset_index())
#seller A | total_orders=3 | total_revenue=450 | average_review=4 | review_count=3

In [ ]:
print("Seller summary shape:",seller_summary_delivered.shape)

print("Duplicate seller IDs:",seller_summary_delivered["seller_id"].duplicated().sum())

In [ ]:
seller_summary_delivered = seller_summary_delivered.merge(sellers,on="seller_id",how="left")

In [ ]:
print("Seller summary rows:",len(seller_summary_delivered))

print("Duplicate seller IDs:",seller_summary_delivered["seller_id"].duplicated().sum())

print("Missing seller state:",seller_summary_delivered["seller_state"].isna().sum())

print("Missing seller city:",seller_summary_delivered["seller_city"].isna().sum())

In [ ]:
print("Minimum orders:",seller_summary_delivered["total_orders"].min())
print("Maximum orders:",seller_summary_delivered["total_orders"].max())
print("Minimum sales value:",seller_summary_delivered["total_revenue"].min())
print("Maximum sales value:",seller_summary_delivered["total_revenue"].max())
print("Minimum review score:",seller_summary_delivered["average_review_score"].min())
print("Maximum review score:",seller_summary_delivered["average_review_score"].max())

In [ ]:
funnel_clean = mql.merge(closed,on="mql_id",how="left")

In [ ]:
print("MQL rows:", len(mql))
print("Funnel rows:", len(funnel_clean))

In [ ]:
converted_sellers = funnel_clean[funnel_clean["seller_id"].notna()].copy()

In [ ]:
print("Converted sellers:",len(converted_sellers))
print("Duplicate seller IDs:",converted_sellers["seller_id"].duplicated().sum())

In [ ]:
seller_analysis = converted_sellers.merge(seller_summary_delivered,on="seller_id",how="left")

In [ ]:
print("Converted sellers:",len(converted_sellers))
print("Seller analysis rows:",len(seller_analysis))
print("Duplicate seller IDs:",seller_analysis["seller_id"].duplicated().sum())

In [ ]:
print("Sellers with delivered performance:",seller_analysis["total_orders"].notna().sum())
print("Sellers without delivered performance:",seller_analysis["total_orders"].isna().sum())

In [ ]:
print(seller_analysis[["total_orders","total_revenue","average_review_score","seller_state"]].isna().sum())

In [ ]:
print("First MQL contact:",mql["first_contact_date"].min())
print("Last MQL contact:",mql["first_contact_date"].max())
print("First won seller:",closed["won_date"].min())
print("Last won seller:",closed["won_date"].max())
print("First order:",orders["order_purchase_timestamp"].min())
print("Last order:",orders["order_purchase_timestamp"].max())

In [ ]:
print("Missing order status after merge:",seller_order_level_all["order_status"].isna().sum())
print("Missing seller state after merge:",seller_order_level_all["seller_state"].isna().sum())
print("Missing review score after merge:",seller_order_level_all["review_score"].isna().sum())

# Part 2 — Analysis Questions 1 to 6
This section uses the cleaned DataFrames already created in Part 1. It does not reload or reclean the raw files.


In [ ]:
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
seller_order_level=seller_order_level_all.copy()
delivered_so=seller_order_level_delivered.copy()
seller_summary=seller_summary_delivered.copy()
seller_summary["avg_order_value"]=seller_summary["total_revenue"]/seller_summary["total_orders"]


# Q1 & Q2

In [ ]:

funnel=funnel_clean.copy()
channel_summary = (funnel.groupby("origin").agg(total_mqls=("mql_id", "count"), closed_deals=("seller_id", "count")).reset_index())

channel_sellers = funnel.dropna(subset=["seller_id"]).merge(seller_summary, on="seller_id", how="left")
channel_perf = ( channel_sellers.groupby("origin").agg(total_orders=("total_orders", "sum"),total_revenue=("total_revenue", "sum"),).reset_index())
channel_df = channel_summary.merge(channel_perf, on="origin", how="left").fillna(0)
channel_df = channel_df.sort_values(by="total_revenue", ascending=False)
fig, ax1 = plt.subplots(figsize=(10, 5))

color = "blue"
ax1.set_xlabel("Acquisition Channel / Origin", fontweight="bold", labelpad=10)
ax1.set_ylabel("Total Revenue (R$)", color=color, fontweight="bold")
bars = ax1.bar( channel_df["origin"],channel_df["total_revenue"], color=color, alpha=0.85,width=0.6,)
ax1.tick_params(axis="y", labelcolor=color)
plt.xticks(rotation=30, ha="right")

ax2 = ax1.twinx()
color = "orange"
ax2.set_ylabel("Total Delivered Orders", color=color, fontweight="bold")
lines = ax2.plot(channel_df["origin"],channel_df["total_orders"],color=color,marker="o", linewidth=2.5,markersize=7,)
ax2.tick_params(axis="y", labelcolor=color)
plt.grid(False)
plt.title("Acquisition Channel Performance: Total Revenue vs. Order Volume",fontsize=13,fontweight="bold",pad=15,)
fig.tight_layout()
plt.show()


# Q3

In [ ]:
seller_revenue=seller_summary_delivered[["seller_id","total_revenue"]].rename(columns={"total_revenue":"total_seller_revenue"})
sr_performance=closed.merge(seller_revenue,on="seller_id",how="left").fillna({"total_seller_revenue":0})
top_srs=(sr_performance.groupby("sr_id")["total_seller_revenue"].sum().reset_index().sort_values(by="total_seller_revenue",ascending=False).head(10))
top_srs["sr_short"]=top_srs["sr_id"].apply(lambda x:str(x)[:8]+"...")
plt.figure(figsize=(10,5))
sns.set_theme(style="whitegrid")
ax=sns.barplot(data=top_srs,x="total_seller_revenue",y="sr_short",palette="Purples_r")
for p in ax.patches:
    width=p.get_width()
    ax.annotate(f" {width:,.0f}",(width,p.get_y()+p.get_height()/2.0),ha="left",va="center",fontsize=9,xytext=(5,0),textcoords="offset points")
plt.title("Employee Selection Quality — Top SRs by Onboarded Seller Revenue",fontsize=13,fontweight="bold",pad=15)
plt.xlabel("Total Delivered Merchandise Revenue",fontweight="bold")
plt.ylabel("Sales Representative ID (SR)",fontweight="bold")
plt.xlim(0,top_srs["total_seller_revenue"].max()*1.18)
plt.tight_layout()
plt.grid(False)
plt.show()


# Q4

In [ ]:
top_10_sellers = (seller_summary.sort_values(by="total_revenue", ascending=False).head(10).copy())

top_10_sellers["seller_label"] = [f"{sid[:8]}..." for i, sid in enumerate(top_10_sellers["seller_id"])]

fig, ax = plt.subplots(figsize=(10, 5.5))
sns.set_theme(style="whitegrid")

sns.barplot(data=top_10_sellers, x="total_revenue", y="seller_label",  palette="Blues_r",ax=ax,)


for p in ax.patches:
    width = p.get_width()
    ax.annotate(
        f" {width:,.0f}",
        (width, p.get_y() + p.get_height() / 2.0),
        ha="left",
        va="center",
        fontsize=9,
        xytext=(5, 0),
        textcoords="offset points",)

ax.set_title("Top 10 Sellers Concentration by Merchandise Revenue", fontsize=13,fontweight="bold",pad=15,)
ax.set_xlabel("Total Merchandise Revenue ", fontweight="bold")
ax.set_ylabel("Seller Rank & ID", fontweight="bold")
ax.set_xlim(0, top_10_sellers["total_revenue"].max() * 1.18)
plt.grid(False)
fig.tight_layout()
plt.show()

# Q5

In [ ]:
seller_flags_data=seller_order_level_delivered.copy()
seller_flags_data["is_late"]=seller_flags_data["order_delivered_customer_date"]>seller_flags_data["order_estimated_delivery_date"]
seller_flags=(seller_flags_data.groupby("seller_id").agg(total_orders=("order_id","nunique"),avg_review_score=("review_score","mean"),late_orders=("is_late","sum")).reset_index())
seller_flags["late_delivery_rate_%"]=(seller_flags["late_orders"]/seller_flags["total_orders"])*100
red_flag_sellers=seller_flags[(seller_flags["total_orders"]>=10)&((seller_flags["avg_review_score"]<3.0)|(seller_flags["late_delivery_rate_%"]>20.0))].sort_values(by="late_delivery_rate_%",ascending=False)
top_red_flags=red_flag_sellers.head(10).copy()
top_red_flags["seller_short"]=top_red_flags["seller_id"].apply(lambda x:str(x)[:8]+"...")
plt.figure(figsize=(10,5))
sns.set_theme(style="whitegrid")
ax=sns.barplot(data=top_red_flags,x="late_delivery_rate_%",y="seller_short",palette="Reds_r")
for p in ax.patches:
    width=p.get_width()
    ax.annotate(f"{width:.1f}% late",(width,p.get_y()+p.get_height()/2.0),ha="left",va="center",fontsize=9,xytext=(5,0),textcoords="offset points")
plt.title("Top Red-Flag Sellers by Late Delivery Rate (Min 10 Orders)",fontsize=13,fontweight="bold",pad=15)
plt.xlabel("Late Delivery Rate (%)",fontweight="bold")
plt.ylabel("Seller ID",fontweight="bold")
plt.xlim(0,top_red_flags["late_delivery_rate_%"].max()*1.18)
plt.tight_layout()
plt.grid(False)
plt.show()


# Q6

In [ ]:
biz_summary = closed.merge(seller_summary, on="seller_id", how="inner")
type_summary = (
    biz_summary.groupby("business_type").agg(
        seller_count=("seller_id", "nunique"),
        total_orders=("total_orders", "sum"),
        total_revenue=("total_revenue", "sum"),
        avg_order_value=("avg_order_value", "mean"),).reset_index())

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 5))

sns.barplot(data=type_summary, x="business_type", y="total_orders", palette="Set2", ax=ax1)
ax1.set_title("Total Delivered Orders", fontweight="bold")
ax1.set_xlabel("Business Model Type", fontweight="bold")
ax1.set_ylabel("Order Count", fontweight="bold")

for p in ax1.patches:
    ax1.annotate(
        f"{int(p.get_height()):,}",
        (p.get_x() + p.get_width() / 2.0, p.get_height()),
        ha="center",
        va="bottom",
        fontsize=10,
        xytext=(0, 3),
        textcoords="offset points",)


sns.barplot(data=type_summary, x="business_type", y="avg_order_value", palette="Set2", ax=ax2,)
ax2.set_title("Average Order Value (AOV)", fontweight="bold")
ax2.set_xlabel("Business Model Type", fontweight="bold")
ax2.set_ylabel("Average Order Value ", fontweight="bold")

for p in ax2.patches:
    ax2.annotate(
        f" {p.get_height():.2f}",
        (p.get_x() + p.get_width() / 2.0, p.get_height()),
        ha="center",
        va="bottom",
        fontsize=10,
        xytext=(0, 3),
        textcoords="offset points",)

plt.suptitle("Manufacturer vs. Reseller Commercial Performance Breakdown",fontsize=14,fontweight="bold",y=1.02,)
fig.tight_layout()
plt.show()

# Part 3 — Analysis Questions 7 to 12
This section continues directly from the same cleaning outputs. The Q10 activity rule is set to **7 days**.


**Q7 — What’s the number of orders for each seller type?**

In [ ]:
seller_analysis.shape

In [ ]:
seller_analysis["business_type"].value_counts(dropna=False)

In [ ]:
seller_type_data = seller_analysis[
    (seller_analysis["business_type"] == "manufacturer") |
    (seller_analysis["business_type"] == "reseller")].copy()

In [ ]:
seller_type_data["business_type"].value_counts()

In [ ]:
orders_by_type = (seller_type_data.groupby("business_type").agg(total_orders=("total_orders", "sum")).reset_index())
orders_by_type

In [ ]:
seller_count_by_type = (seller_type_data.groupby("business_type").agg(seller_count=("seller_id", "nunique")).reset_index())
seller_count_by_type

In [ ]:
seller_type_summary = (seller_type_data.groupby("business_type").agg(seller_count=("seller_id", "nunique"),total_orders=("total_orders", "sum")).reset_index())
seller_type_summary

In [ ]:
plt.bar(seller_type_summary["business_type"],seller_type_summary["total_orders"])

plt.xlabel("Seller Type")
plt.ylabel("Number of Delivered Orders")
plt.title("Delivered Orders by Seller Type")

plt.tight_layout()
plt.show()

***The resellers made 3,881 orders that were delivered, while only 577 orders made by manufacturers were delivered in the available transaction period.***

***Q8 — Are top sellers successful because of price or customer satisfaction?***

In [ ]:
seller_success = seller_summary_delivered.copy()

In [ ]:
# Calculate average item price for each seller

seller_price = (seller_order_level_delivered.groupby("seller_id").agg(total_sales=("order_value", "sum"),total_items=("item_count", "sum")).reset_index())
seller_price["average_item_price"] = (seller_price["total_sales"] / seller_price["total_items"])

In [ ]:
# Add average item price to seller performance data

seller_success = seller_summary_delivered.merge(seller_price[["seller_id", "average_item_price"]],on="seller_id",how="left")

In [ ]:
# Top 10 sellers by delivered sales value

top_10_sellers = (seller_success.sort_values("total_revenue", ascending=False).head(10))
other_sellers = seller_success[~seller_success["seller_id"].isin(top_10_sellers["seller_id"])]

In [ ]:
top_price = top_10_sellers["average_item_price"].mean()
other_price = other_sellers["average_item_price"].mean()





plt.bar(
    ["Top 10 Sellers", "Other Sellers"],
    [top_price, other_price])

plt.ylabel("Average Item Price")
plt.title("Average Item Price: Top Sellers vs Other Sellers")

plt.show()

In [ ]:
top_review=top_10_sellers["average_review_score"].mean()
other_review=other_sellers["average_review_score"].mean()





plt.bar(["Top 10 Sellers","Other Sellers"],[top_review,other_review])
plt.ylabel("Average Review Score")
plt.title("Customer Satisfaction: Top Sellers vs Other Sellers")
plt.ylim(0,5)
plt.show()

***It seems like top sellers perform better due to higher prices of items instead of higher customer satisfaction. The average price of an item of top sellers is 223.75 compared to 177.50 for other sellers, while the average review score is 4.11 compared to 4.18***

***Q9 — Where are the top sellers located?***

In [ ]:
top_10_sellers[["seller_id","seller_city","seller_state","total_revenue"]]

In [ ]:
top_seller_states=top_10_sellers["seller_state"].value_counts().reset_index()
top_seller_states.columns=["state","top_seller_count"]
top_seller_states

In [ ]:
top_10_sellers=top_10_sellers.sort_values("total_revenue",ascending=False).reset_index(drop=True)
top_10_sellers["seller_rank"]=["Seller "+str(i) for i in range(1,11)]
colors=["steelblue" if state=="SP" else "orange" for state in top_10_sellers["seller_state"]]
plt.figure(figsize=(10,5))
plt.bar(top_10_sellers["seller_rank"],top_10_sellers["total_revenue"],color=colors)
plt.xlabel("Top Sellers")
plt.ylabel("Total Revenue")
plt.title("Top 10 Sellers by State")
plt.xticks(rotation=45)
plt.show()

***Most of the top sellers operate in SP. Out of the top 10 sellers, nine of them are from SP and only one is from BA.***

***Q10 — How many sellers do we have, and are all of them active?***

In [ ]:
last_order_date=seller_order_level_all["order_purchase_timestamp"].max()
seller_last_order=seller_order_level_all.groupby("seller_id")["order_purchase_timestamp"].max().reset_index()


In [ ]:
seller_activity=sellers[["seller_id"]].merge(seller_last_order,on="seller_id",how="left")
seller_activity["days_since_last_order"]=(last_order_date-seller_activity["order_purchase_timestamp"]).dt.days
seller_activity["active"]=seller_activity["days_since_last_order"]<=7
total_sellers=seller_activity["seller_id"].nunique()
active_sellers=seller_activity["active"].sum()
inactive_sellers=total_sellers-active_sellers


In [ ]:
print("Total sellers:",total_sellers)
print("Active sellers:",active_sellers)
print("Inactive sellers:",inactive_sellers)
print("Last order date:",last_order_date)
print(f"Answer: There are {total_sellers:,} sellers in total. Using the 7-day activity definition, {active_sellers:,} are active and {inactive_sellers:,} are inactive.")


In [ ]:
activity_counts=seller_activity["active"].value_counts().reset_index()
activity_counts.columns=["active","seller_count"]
activity_counts["status"]=activity_counts["active"].map({True:"Active",False:"Inactive"})
plt.bar(activity_counts["status"],activity_counts["seller_count"])
plt.xlabel("Seller Status")
plt.ylabel("Number of Sellers")
plt.title("Active vs Inactive Sellers")
plt.show()

***Q10 Answer — There are 3,095 sellers in total. For this project, an active seller is defined as a seller whose most recent recorded order occurred within 7 days of the final order date in the dataset. The exact active and inactive counts are generated by the code above.***


***Q11 — Which states have customers but zero sellers?***

In [ ]:
customers_by_state=customers.groupby("customer_state")["customer_unique_id"].nunique().reset_index()
customers_by_state.columns=["state","customer_count"]


In [ ]:
sellers_by_state=sellers.groupby("seller_state")["seller_id"].nunique().reset_index()
sellers_by_state.columns=["state","seller_count"]


In [ ]:
state_coverage=customers_by_state.merge(sellers_by_state,on="state",how="left")


In [ ]:
state_coverage["seller_count"]=state_coverage["seller_count"].fillna(0)
states_no_sellers=state_coverage[state_coverage["seller_count"]==0]
states_no_sellers

In [ ]:
plt.figure(figsize=(7,5))
plt.bar(states_no_sellers["state"],states_no_sellers["customer_count"])
plt.xlabel("State")
plt.ylabel("Number of Unique Customers")
plt.title("States with Customers but Zero Sellers")
plt.show()

***There are four states where there are some customers, but no sellers: AL, AP, RR, and TO. The state AL has the highest number of customers with 401 customers, then TO with 273, AP with 67, and RR with 45 customers.***

ٍٍ***Q12 — Do we have a good seller-to-customer ratio?***

In [ ]:
state_coverage["sellers_per_1000_customers"]=(state_coverage["seller_count"]/state_coverage["customer_count"])*1000
state_ratio=state_coverage.sort_values("sellers_per_1000_customers",ascending=True)
state_ratio[["state","customer_count","seller_count","sellers_per_1000_customers"]]

In [ ]:
plt.figure(figsize=(10,8))
plt.barh(state_ratio["state"],state_ratio["sellers_per_1000_customers"])
plt.xlabel("Sellers per 1,000 Customers")
plt.ylabel("State")
plt.title("Seller-to-Customer Ratio by State")
plt.show()

***The ratio between seller and customer is uneven in different states. PR has the highest seller coverage of approximately 71 sellers per 1,000 customers, followed by SC and SP. AL, AP, RR, and TO states have customers, but no sellers.***

# Part 4 — Additional Analysis Questions
This section uses the same cleaned DataFrames from Part 1. No separate cleaned CSV files are required.

**Important definition:** In this section, `active_seller` is used as an **activation** flag for signed sellers (has recorded delivered-order performance). This is different from Q10, where 'active' means the seller's last recorded order was within 7 days of the end of the dataset.


In [ ]:
seller_order_level=seller_order_level_all.copy()
seller_analysis["revenue_per_order"]=seller_analysis["total_revenue"]/seller_analysis["total_orders"]
first_sale=seller_order_level_all.groupby("seller_id")["order_purchase_timestamp"].min().reset_index(name="first_sale_date")
seller_analysis=seller_analysis.merge(first_sale,on="seller_id",how="left")
seller_analysis["days_to_first_sale"]=(seller_analysis["first_sale_date"]-seller_analysis["won_date"]).dt.days


# Olist Seller Evaluation & Channel Analysis
## Assigned Analysis Questions
Which employees are hiring the best vs. worst sellers?

Provide specific, data-driven recommendations.

Explore population density and demographics in states with no sellers.

Compare average revenue per order between manufacturers and resellers.

Assess whether seller location affects delivery time.

Which lead channel (e.g., Organic Search, Paid Search, Social) brings sellers with the highest retention and order volume?

What percentage of signed sellers never list a product or make their first sale within 30/60 days of onboarding?

## 1. Which employees are hiring the best vs. worst sellers?

In [ ]:
# Activation flag for signed sellers in this section:
# True = seller has recorded delivered-order performance
seller_analysis["active_seller"]=seller_analysis["total_orders"].notna()
seller_analysis["active_seller"].value_counts()


In [ ]:
#SR performance table
sr_summary = seller_analysis.groupby("sr_id").agg(
    sellers_hired=("seller_id", "count"),
    active_sellers=("active_seller", "sum"),
    average_orders=("total_orders", "mean"),
    average_revenue=("total_revenue", "mean"),
    average_review=("average_review_score", "mean")
).reset_index()

In [ ]:
#activation rate
sr_summary["active_rate"] = (sr_summary["active_sellers"] /sr_summary["sellers_hired"]) * 100

In [ ]:
sr_summary

In [ ]:
#filter out very small employee samples
sr_summary_fair = sr_summary[sr_summary["sellers_hired"] >= 10].copy()

print("SRs included:", len(sr_summary_fair))

In [ ]:
#strongest activation rates
sr_summary_fair.sort_values("active_rate",ascending=False)

In [ ]:
#strongest average order volume
sr_summary_fair.sort_values("average_orders",ascending=False)

In [ ]:
#strongest average revenue
sr_summary_fair.sort_values("average_revenue",ascending=False)

In [ ]:
#separate analysis copy
sr_analysis_data = seller_analysis.copy()

In [ ]:
#treat sellers with no recorded transactions as zero for this specific employee-performance calculation
sr_analysis_data["total_orders_for_employee"] = (sr_analysis_data["total_orders"].fillna(0))

sr_analysis_data["total_revenue_for_employee"] = (sr_analysis_data["total_revenue"].fillna(0))

In [ ]:
#overall SR table
sr_overall = sr_analysis_data.groupby("sr_id").agg(
    sellers_hired=("seller_id", "count"),
    active_sellers=("active_seller", "sum"),
    total_orders=("total_orders_for_employee", "sum"),
    total_revenue=("total_revenue_for_employee", "sum"),
    average_review=("average_review_score", "mean")
).reset_index()

In [ ]:
sr_overall["active_rate"] = (sr_overall["active_sellers"] /sr_overall["sellers_hired"]) * 100

In [ ]:
#productivity per seller hired
sr_overall["orders_per_hired_seller"] = (sr_overall["total_orders"] /sr_overall["sellers_hired"])

sr_overall["revenue_per_hired_seller"] = (sr_overall["total_revenue"] /sr_overall["sellers_hired"])

In [ ]:
sr_overall_fair = sr_overall[sr_overall["sellers_hired"] >= 10].copy()

In [ ]:
sr_overall_fair.sort_values("revenue_per_hired_seller",ascending=False)

In [ ]:
sr_overall_fair.sort_values("orders_per_hired_seller",ascending=False)

In [ ]:
#SDR side
sdr_overall = sr_analysis_data.groupby("sdr_id").agg(
    sellers_handled=("seller_id", "count"),
    active_sellers=("active_seller", "sum"),
    total_orders=("total_orders_for_employee", "sum"),
    total_revenue=("total_revenue_for_employee", "sum"),
    average_review=("average_review_score", "mean")
).reset_index()

In [ ]:
sdr_overall["active_rate"] = (sdr_overall["active_sellers"] /sdr_overall["sellers_handled"]) * 100

In [ ]:
sdr_overall["orders_per_seller"] = (sdr_overall["total_orders"] /sdr_overall["sellers_handled"])
sdr_overall["revenue_per_seller"] = (sdr_overall["total_revenue"] /sdr_overall["sellers_handled"])

In [ ]:
sdr_overall_fair = sdr_overall[sdr_overall["sellers_handled"] >= 10].copy()

In [ ]:
#Filter small samples
sdr_overall_fair = sdr_overall[sdr_overall["sellers_handled"] >= 10].copy()

In [ ]:
sdr_overall_fair.sort_values("active_rate",ascending=False)

In [ ]:
sdr_overall_fair.sort_values("orders_per_seller",ascending=False)

In [ ]:
sdr_overall_fair.sort_values("revenue_per_seller",ascending=False)

## 2. Compare average revenue per order between manufacturers and resellers

In [ ]:
seller_analysis["business_type"].value_counts(dropna=False)

In [ ]:
#manufacturers and resellers with transactions
business_type_active = seller_analysis[(seller_analysis["business_type"].isin(["manufacturer", "reseller"]))&(seller_analysis["revenue_per_order"].notna())
].copy()

In [ ]:
business_type_active["business_type"].value_counts()

In [ ]:
#average revenue per order
business_type_comparison = business_type_active.groupby("business_type")["revenue_per_order"].mean()
business_type_comparison

In [ ]:
#median
business_type_active.groupby("business_type")["revenue_per_order"].median()

In [ ]:
#Adding order volume and seller count for context
business_type_summary = business_type_active.groupby("business_type").agg(
    sellers=("seller_id", "count"),
    average_orders=("total_orders", "mean"),
    average_revenue_per_order=("revenue_per_order", "mean"),
    median_revenue_per_order=("revenue_per_order", "median"),
    average_review=("average_review_score", "mean")
).reset_index()

business_type_summary

## 3. Which lead channel brings sellers with the highest retention and order volume?

In [ ]:
#Build channel summary
channel_summary = seller_analysis.groupby("origin").agg(
    signed_sellers=("seller_id", "count"),
    active_sellers=("active_seller", "sum"),
    average_orders=("total_orders", "mean"),
    average_revenue=("total_revenue", "mean"),
    average_review=("average_review_score", "mean")
).reset_index()

In [ ]:
channel_summary["active_rate"] = (channel_summary["active_sellers"] /channel_summary["signed_sellers"]) * 100

In [ ]:
channel_summary

In [ ]:
#Rank by active rate
channel_summary.sort_values("active_rate",ascending=False)

In [ ]:
#Rank by order volume
channel_summary.sort_values("average_orders",ascending=False)

In [ ]:
channel_summary[["origin", "signed_sellers", "active_sellers", "active_rate", "average_orders"]]

## 4. What percentage of signed sellers never list a product or make their first sale within 30/60 days of onboarding?

In [ ]:
print("Total signed sellers:", len(seller_analysis))
print("Sellers with a recorded first sale:",seller_analysis["first_sale_date"].notna().sum())
print("Sellers with no recorded first sale:",seller_analysis["first_sale_date"].isna().sum())

In [ ]:
not_sold_30 = (seller_analysis["first_sale_date"].isna()|(seller_analysis["days_to_first_sale"] > 30))

print("Not sold within 30 days:", not_sold_30.sum())

print("Percentage:",round(not_sold_30.mean() * 100, 2),"%")

In [ ]:
not_sold_60 = (seller_analysis["first_sale_date"].isna()|(seller_analysis["days_to_first_sale"] > 60))
print("Not sold within 60 days:", not_sold_60.sum())

print("Percentage:",round(not_sold_60.mean() * 100, 2),"%")

In [ ]:
sold_30 = (seller_analysis["first_sale_date"].notna()&(seller_analysis["days_to_first_sale"] <= 30))
sold_60 = (seller_analysis["first_sale_date"].notna()&(seller_analysis["days_to_first_sale"] <= 60))

print("Sold within 30 days:", sold_30.sum())
print("Sold within 60 days:", sold_60.sum())

print("Percentage sold within 30 days:",round(sold_30.mean() * 100, 2),"%")
print("Percentage sold within 60 days:",round(sold_60.mean() * 100, 2),"%")

## 5. Assess whether seller location affects delivery time

In [ ]:
#seller-delivery data
delivery_analysis = seller_order_level.merge(delivered_orders[["order_id", "delivery_days"]],on="order_id",how="inner")

In [ ]:
print("Delivery analysis rows:", delivery_analysis.shape)

In [ ]:
#Checkiing seller state is present
print(delivery_analysis["seller_state"].isna().sum())

In [ ]:
#Calculating delivery time by seller state
delivery_by_state = delivery_analysis.groupby("seller_state").agg(
    orders=("order_id", "nunique"),
    average_delivery_days=("delivery_days", "mean"),
    median_delivery_days=("delivery_days", "median")
).reset_index()

In [ ]:
delivery_by_state.sort_values("average_delivery_days",ascending=False)

In [ ]:
#Avoiding tiny state samples
delivery_by_state_fair = delivery_by_state[delivery_by_state["orders"] >= 30].copy()

In [ ]:
delivery_by_state_fair.sort_values("average_delivery_days",ascending=False)

### 6. Explore population density and demographics in states with no sellers

In [ ]:
# customers and sellers already come from Part 1 cleaning
print("Customers:",customers.shape)
print("Sellers:",sellers.shape)


In [ ]:
#unique states
customer_states = customers["customer_state"].unique()
seller_states = sellers["seller_state"].unique()

print("Customer states:")
print(customer_states)

print("\nSeller states:")
print(seller_states)

In [ ]:
#states that appear in customers but not sellers
states_with_customers_no_sellers = [state
    for state in customer_states
    if state not in seller_states
]

print("States with customers but no sellers:",states_with_customers_no_sellers)

In [ ]:
#unique customers in states without sellers 
customers_no_seller_states = customers[customers["customer_state"].isin(["AP", "AL", "TO", "RR"])]

customers_no_seller_states.groupby("customer_state")["customer_unique_id"].nunique().sort_values(ascending=False)

In [ ]:
#share of all customers
customer_counts_no_sellers = customers_no_seller_states.groupby("customer_state")["customer_unique_id"].nunique()
total_unique_customers = customers["customer_unique_id"].nunique()
customer_share_no_sellers = (customer_counts_no_sellers /total_unique_customers) * 100
customer_share_no_sellers.sort_values(ascending=False)

In [ ]:
#demand
customer_activity_no_sellers = customers_no_seller_states.groupby("customer_state").agg(
    unique_customers=("customer_unique_id", "nunique"),
    customer_records=("customer_id", "count")
).reset_index()

customer_activity_no_sellers

In [ ]:
#average customer records per unique customer
customer_activity_no_sellers["records_per_customer"] = (customer_activity_no_sellers["customer_records"]/customer_activity_no_sellers["unique_customers"])
customer_activity_no_sellers.sort_values("unique_customers",ascending=False)

## 7. Provide specific, data-driven recommendations

## Recommendation Framework
Use the calculated results above to support the final recommendations:
- Prioritize acquisition channels with stronger activation and order volume.
- Strengthen seller onboarding during the first 30–60 days.
- Benchmark stronger SR/SDR performance and support lower-performing employees.
- Use separate benchmarks for manufacturers and resellers.
- Investigate seller acquisition in customer states with zero sellers, especially the states with the largest customer bases.
- Investigate logistics in states with slower delivery performance.
- Improve tracking for leads classified as `unknown`.
